tools/hsdbg/hsdbg.py

```diff
@@ -0,0 +1,323 @@
+    return frozenset(attrList)
+
+  def parseDotEdge(self, i, line):
+    """Parses a graphviz file edge. An edge has the format:
+                "from" -> "to" [optional attributes] // custom attributes
+        """
+    src, rest = line.split("->")[0].strip(), line.split("->")[1].strip()
+
+    rest = rest.split("//")
+    customAttributes = frozenset()
+    if len(rest) > 1:
+      customAttributes = self.parseCustomAttributes(rest[1].strip())
+    rest = rest[0].strip()
+    dstSplit = rest.split("[")
+    dst = dstSplit[0]
+    src = src.replace('"', '').strip()
+    dst = dst.replace('"', '').strip()
+    # split [.* from dst and also return the second part
+    attributes = frozenset()
+    if len(dstSplit) > 1:
+      attrStr = dstSplit[1].replace(']', '')
+      # attributes is a list of key value pairs with '=' in between
+      # and separated by spaces. Parse it as a list of pairs
+      attributes = [x.strip() for x in attrStr.split(" ")]
+      attributes = [x.split("=") for x in attributes if len(x) > 0]
+      attributes = frozenset({
+          (x[0].replace("\"", ""), x[1].replace("\"", "")) for x in attributes
+      })
+    edge = DotEdge(src, dst, attributes, i, customAttributes)
+    self.edgeToLine[edge] = i
+    self.lineToEdge[i] = edge
+    self.edges.add(edge)
+
+  def parseDot(self, filename):
+    with open(filename, 'r') as f:
+      self.rawDot = f.readlines()
+
+    # Create edge to line mapping
+    for i, line in enumerate(self.rawDot):
+      if "->" in line:
+        self.parseDotEdge(i, line)
+
+  def dump(self, path):
+    """ write the modified dot file to path."""
+    with open(path, "w") as f:
+      f.write("".join(self.modifiedDot))
+
+
+class HSDbg():
+
+  def __init__(self, target, dotFile, vcdFile):
+    # Factory for building signal bundles and modules.
+    self.target = target
+
+    # Maintain the source code of the dot file. Since there aren't any good
+    # python libraries for in-place modification of
+    self.dotFile = DotFile(dotFile)
+    self.workingImagePath = self.setupWorkingImage(dotFile)
+    # The bundles which have been successfully mapped to a set of
+    # VCD signals and the corresponding dot edges.
+    self.bundles = []
+
+    self.vcd = None
+    self.parseVCD(vcdFile)
+    self.cycle = self.beginTime()
+    self.resolveDotToVCD()
+
+    # Create the graph.
+    self.updateCurrentCycle(self.cycle)
+
+  def resolveDotToVCD(self):
+    """ This function uses the edges gathered from the .dot file to try and
+            create an association with handshake bundles in the vcd file. Be aware
+            that quite a few assumptions are placed on the format of both the vcd
+            file and the dot file.
+        """
+    # Step 1: infer top module name
+    top = "TOP"
+
+    hier = resolveVCDHierarchy(self.vcd)
+    if len(hier) != 1 or "TOP" not in hier:
+      raise Exception("Expected \"TOP\" as top-level in vcd hierarchy")
+    topHier = hier["TOP"]
```

Comment:
I think you can relax this to just enforce there is only one top-level, and take that, similar to how you find dutHier below. I generated a VCD with a different top name, and this works:

```suggestion
    if len(hier) != 1:
      raise Exception("Expected one top-level in vcd hierarchy")
    top = list(hier.keys())[0]
    topHier = hier[top]
```